UnitTests/GTMSessionFetcherChunkedUploadTest.m

```diff
@@ -809,6 +809,130 @@ - (void)testBigFileURLSingleChunkedUploadFetchRetry {
+  [self assertCallbacksReleasedForFetcher:fetcher];
+
+  WAIT_FOR_START_STOP_NOTIFICATION_EXPECTATIONS();
+
+  // Check that we uploaded the expected chunks.
+  NSArray *expectedCommands = @[
+    @"query", @"upload", @"query", @"upload", @"query", @"upload", @"query", @"upload", @"query",
+    @"upload", @"query", @"upload", @"query", @"upload", @"query"
+  ];
+  NSArray *expectedOffsets = @[ @0, @0, @0, @0, @0, @0, @0, @0, @0, @0, @0, @0, @0, @0, @0 ];
+  NSArray *expectedLengths =
+      @[ @0, @5000, @0, @5000, @0, @5000, @0, @5000, @0, @5000, @0, @5000, @0, @5000, @0 ];
+  XCTAssertEqualObjects(fnctr.uploadChunkCommands, expectedCommands);
+  XCTAssertEqualObjects(fnctr.uploadChunkOffsets, expectedOffsets);
+  XCTAssertEqualObjects(fnctr.uploadChunkLengths, expectedLengths);
+
+  XCTAssertEqual(fnctr.fetchStarted, 15);
+  XCTAssertEqual(fnctr.fetchStopped, 15);
+  XCTAssertEqual(fnctr.uploadChunkFetchStarted, 15);
+  XCTAssertEqual(fnctr.uploadChunkFetchStopped, 15);
+  XCTAssertEqual(fnctr.retryDelayStarted, 0);
+  XCTAssertEqual(fnctr.retryDelayStopped, 0);
+  XCTAssertEqual(fnctr.uploadLocationObtained, 0);
+
+  [self removeTemporaryFileURL:bigFileURL];
+}
+
+// Forces the server to return a 503 every time there is an upload, cancels the upload in the middle
+// of a backoff, and makes sure that the timer has been killed.
+- (void)testBigFileURLSingleChunkedUploadFetchLimitedRetryWithCancel {
+  CREATE_START_STOP_NOTIFICATION_EXPECTATIONS(4, 4);
+  FetcherNotificationsCounter *fnctr = [[FetcherNotificationsCounter alloc] init];
+
+  NSURL *bigFileURL = [self bigFileToUploadURLWithBaseName:NSStringFromSelector(_cmd)];
+  NSString *filename = [NSString stringWithFormat:@"gettysburgaddress.txt.upload?uploadStatus=503"];
+  NSURL *uploadLocationURL = [_testServer localURLForFile:filename];
+
+  GTMSessionUploadFetcher *fetcher =
+      [GTMSessionUploadFetcher uploadFetcherWithLocation:uploadLocationURL
+                                          uploadMIMEType:@"text/plain"
+                                               chunkSize:5000
+                                          fetcherService:_service];
+
+  __weak typeof(fetcher) weakFetcher = fetcher;
+  fetcher.uploadFileURL = bigFileURL;
+  fetcher.useBackgroundSession = NO;
+  fetcher.allowLocalhostRequest = YES;
+
+  XCTestExpectation *expectation = [self expectationWithDescription:@"completion handler"];
+  fetcher.retryEnabled = YES;
+
+  fetcher.cancellationHandler = ^(GTMSessionFetcher *_Nullable sessionFetcher,
+                                  NSData *_Nullable data, NSError *_Nullable error) {
+    __strong typeof(fetcher) strongFetcher = weakFetcher;
+    XCTAssertNil(strongFetcher.uploadRetryTimer);
+    XCTAssertNotNil(sessionFetcher);
+    [expectation fulfill];
+  };
+  fetcher.backoffBlock = ^() {
+    __strong typeof(fetcher) strongFetcher = weakFetcher;
+    [strongFetcher performSelector:@selector(stopFetching) withObject:nil afterDelay:0.0];
```

Comment:
To call a single method, the weak->strong conversion here doesn't need to be explicit:

```
[weakFetcher performSelector:@selector(stopFetching) withObject:nil afterDelay:0.0];
```